drivers/saul/pwm_saul.c

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright (C) 2015 Freie Universität Berlin
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     drivers_saul
+ * @{
+ *
+ * @file
+ * @brief       SAUL wrapper for PWM pins
+ *
+ * @author      Christian Amsüss <chrysn@fsfe.org>
+ *
+ * @}
+ */
+
+#include "saul.h"
+#include "phydat.h"
+#include "periph/pwm.h"
+#include "saul/periph.h"
+#include "bitarithm.h"
+
+/* Read operations are not supported through the PWM device API */
+static int no_read(const void *dev, phydat_t *state)
+{
+    (void)dev;
+    (void)state;
+    return -ENOTSUP;
+}
+
+/**
+ * Find factor and shiftback such that for each value entry in the phydat, the
+ * resulting PWM duty cycle would be (value * factor) >> shiftback.
+ */
+static int extract_scaling(const phydat_t *state, int *factor, int *shiftback)
+{
+    if (state->scale != 0) {
+        return -ECANCELED;
+    }
+
+    /** Number of bits i by which we can shift the calculation (value * (255 << i)/100) >> i
+     * to get a better result than value * 2 (which would otherwise happen in integers) */
+    int shift100 = bitarithm_msb(INT_MAX) - bitarithm_msb(saul_pwm_resolution);
+
+    switch (state->unit) {
+    case UNIT_UNDEF:
+    case UNIT_NONE:
+        *factor = 1;
+        *shiftback = 0;
+        break;
+    case UNIT_BOOL:
+        *factor = saul_pwm_resolution;
+        *shiftback = 0;
+        break;
+    case UNIT_PERCENT:
+        *factor = ((int)saul_pwm_resolution << shift100) / 100;
+        *shiftback = shift100;
+        break;
+    default:
+        return -ECANCELED;
+    }
+
+    return 0;
+}
+
+static inline void setchan(const saul_pwm_channel_t *chan, uint16_t value)
+{
+    pwm_set(chan->dev,
+            chan->channel,
+            (chan->flags & SAUL_PWM_INVERTED) ? saul_pwm_resolution - value : value);
+}
+
+static int write_dimmer(const void *dev, phydat_t *state)
+{
+    const saul_pwm_dimmer_params_t *p = (const saul_pwm_dimmer_params_t *)dev;
+
+    int factor, shiftback;
+
+    int err = extract_scaling(state, &factor, &shiftback);
+    if (err < 0)
+        return err;
```

Comment:
```suggestion
    if (err < 0) {
        return err;
    }
```

Reply:
Oh, right, there's a difference to the Linux convention. Thanks, will be in right away with the rest.